### ISSUE TITLE
Calling fromStringArray to populate doesn't set dirty flag

### ISSUE THREAD

**Issue by gerrywastaken** [calcinai/xero-php#40]
I wasn't sure of the best way of solving this. If I populate a new remote object using fromStringArray() and then call save on it, nothing happens. After looking at the code it is because the dirty flag isn't set so the first check during a save fails and the function just exits without any reason as to why.

I initially thought of setting the dirty flag during the fromStringArray, but as you are using fromStringArray() to populate objects after a read from Xero, this isn't appropriate. Any ideas on this?

**Comment by calcinai** [comment]
Yeah, it was originally intended as a method for specifically saturating the objects from Xero.  By default, I guess there's no harm in setting it dirty by default when it's called, then in the 10-odd places it's called, follow it with a ```->setClean()```

Thinking about it, when you're putting data into an object it naturally suggests that it would be dirty.. What do you reckon?

**Comment by gerrywastaken** [comment]
Another option I was thinking of was adding a second param for the dirty state with a default of true, then in your calls you would just pass false.

> Thinking about it, when you're putting data into an object it naturally suggests that it would be dirty... What do you reckon?

Under a normal situation yes, but not in your case. Maybe there needs to be a private and public version of the method? With public always being dirty and private always being clean....

**Comment by calcinai** [comment]
I just did a bit more digging - half the time it is used (Application.php [L184](https://github.com/calcinai/xero-php/blob/master/src/XeroPHP/Application.php#L184), [L236](https://github.com/calcinai/xero-php/blob/master/src/XeroPHP/Application.php#L236), [L275](https://github.com/calcinai/xero-php/blob/master/src/XeroPHP/Application.php#L275)), it is actually assuming a dirty object anyway, so would make no difference if it set it dirty.

If the ```fromStringArray()``` method called the ```propertyUpdated()``` method internally, it'd solve this problem entirely I think.

**Comment by gerrywastaken** [comment]
Well you understand the code far better than I do... heck I didn't even know that method existed. However won't it always set an object to dirty if it wasn't previously populated... due to the check assuming dirty if the value wasn't set:

`if(!isset($this->_data[$property]) || $this->_data[$property] !== $value){`

But as you said, this doesn't really matter if everywhere that the internal population is done, you are setting the object to be clean afterwards.

**Comment by calcinai** [comment]
Hmm. I'll test it out.  Last thing I want is to introduce something that's not BC.

**Comment by gerrywastaken** [comment]
Cool. Although I think the usage is still quite low at the moment, and I'll bet the percent that discovered that function is a lot lower. The ease of the change is also quite low from the consumer end.

**Comment by calcinai** [comment]
Still not sure of the best way to solve this.  Feels _dirty_ (ha) to use that ```propertyUpdated()``` method like this.

Main issue being is that the object itself isn't flagged as dirty, the individual properties are.  The ```fromStringArray()``` method is a function on the object, not the property.  See what I mean?

**Comment by gerrywastaken** [comment]
Yeah I'd noticed that.... how about

```php
$this->setDirty('_data');
```

**Comment by calcinai** [comment]
I still can't work out a nice way to implement this!  Maybe it should be something that's considered in the 2.x overhaul (if you've got a work around so far?)

**Comment by calcinai** [comment]
I've done some testing with that, but it doesn't really solve the issue if I'm not calling ->setClean() after each time it's used internally, so may as well go with the ```->propertyUdated()``` > ```->setClean()``` idea..

**Comment by wildflower** [comment]
Hi, thanks for your work on this library.
(this might be in the wrong place)

I've managed to retrieve and store Contacts from Xero as a JSON string in MySQL and now I would like to recreate the Contact object from the stored JSON string.
I managed via a fudge below but I'm thinking there's probably a nicer way using fromStringArray ?:

	public function getLoadcontact()
	{		
		$realContact = new Contact();
		$properties = $realContact->getProperties();
		$contactobject = $this->getContactobject();
		foreach($properties as $key => $value){
			//echo "key is: $key and value is: ";
			if(isset($contactobject->$key)){
				if($key == 'Addresses'){
					$action = 'addAddress';
					$values = new Address();
				}elseif($key == 'Phones'){
					$action = 'addPhone';
					$values = new Phone();
				}elseif($key == 'UpdatedDateUTC'){
						$values = \DateTime::createFromFormat('ISO8601',$contactobject->$key);
				}elseif($key == 'HasAttachments'){
					$action = 'setHasAttachment';
					$values = $contactobject->$key;
				}
				else{
				$action = 'set'.$key;			
				$values = $contactobject->$key;
				}				
				$realContact->$action($values);	
			}			
		}
		return $realContact;		
	}

**Comment by calcinai** [comment]
Yeah, I'd imagine you can use that function to do that.  Be cautious though, as it's designed to map with whats required fro the Xero API so may change.

I wonder if simply using ```serialize()``` would be enough - is your database only going to be read by PHP?  As ugly as it sounds, it might be cleaner.

**Comment by Arhidiham** [comment]
Hello @calcinai, wonderful SDK you have been working on.

Since your discussion this method `fromStringArray()` has already been updated with a second parameter. Since we have established that this function allows for change we can just add a third parameter like this:
`fromStringArray($input_array, $replace_data = false, $is_dirty = false)`

This will allow all your backwards compatibility to be fine and anyone that wants to use this method for ease of use can still use it to setup objects and quickly save them without the need for excessive boilerplate like this:
```
$invoice = new \XeroPHP\Models\Accounting\Invoice();
$invoice->setContact()
    ->setLineItem()
    ->setType();
// etc
```

My suggested change would probably be around where you set:
`$this->_data[$property]` to a new value and then simply call:
`$this->setDirty($property)`

This will ensure that your logic for `$replace_data` would still function as expected as well.
Do you want me to create a fork and request for change to review?

All that said do you have a preferred alternative to setting up objects quickly right before saving?
